### ISSUE TITLE
Errors with "cargo test"

### ISSUE THREAD

**Issue by palfrey** [fleabitdev/glsp#12]
Running from current master HEAD (86386a72d6202e230fc6a3754cd32aa4e9af9fe2) and doing "cargo +nightly test" I get

```
failures:

---- src/lib.rs - compile (line 19) stdout ----
error: cannot find macro `compile` in this scope
 --> src/lib.rs:20:1
  |
3 | compile!["scripts/first.glsp", "scripts/second.glsp"]
  | ^^^^^^^

error: aborting due to previous error

Couldn't compile the test.
---- src/lib.rs - eval (line 82) stdout ----
error: cannot find macro `eval` in this scope
 --> src/lib.rs:87:19
  |
7 | let result: Val = eval!(r#"
  |                   ^^^^

error[E0412]: cannot find type `Val` in this scope
 --> src/lib.rs:87:13
  |
7 | let result: Val = eval!(r#"
  |             ^^^ not found in this scope
  |
help: possible candidate is found in another module, you can import it into scope
  |
2 | use glsp::Val;
  |

error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
  --> src/lib.rs:87:19
   |
2  |  / fn main() {
3  |  | let input = 100_i64;
4  |  | let mut output = "hello".to_string();
5  |  | let assigned_but_not_read: f32;
6  |  |
7  |  | let result: Val = eval!(r#"
   |  |___________________^
8  | ||     (= ~output (str ~input))
9  | ||     (= ~assigned_but_not_read 100.0)
10 | || "#)?;
   | ||____^ cannot use the `?` operator in a function that returns `()`
11 |  | }
   |  |_- this function should return `Result` or `Option` to accept `?`
   |
   = help: the trait `std::ops::Try` is not implemented for `()`
   = note: required by `std::ops::Try::from_error`

error: aborting due to 3 previous errors

Some errors have detailed explanations: E0277, E0412.
For more information about an error, try `rustc --explain E0277`.
Couldn't compile the test.
---- src/lib.rs - eval (line 69) stdout ----
error: cannot find macro `eval` in this scope
 --> src/lib.rs:70:25
  |
3 | let result: Root<Arr> = eval!(r#"
  |                         ^^^^

error[E0412]: cannot find type `Root` in this scope
 --> src/lib.rs:70:13
  |
3 | let result: Root<Arr> = eval!(r#"
  |             ^^^^ not found in this scope
  |
help: possible candidate is found in another module, you can import it into scope
  |
2 | use glsp::Root;
  |

error[E0412]: cannot find type `Arr` in this scope
 --> src/lib.rs:70:18
  |
3 | let result: Root<Arr> = eval!(r#"
  |                  ^^^ not found in this scope
  |
help: possible candidate is found in another module, you can import it into scope
  |
2 | use glsp::Arr;
  |

error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
 --> src/lib.rs:70:25
  |
2 |  / fn main() {
3 |  | let result: Root<Arr> = eval!(r#"
  |  |_________________________^
4 | ||     (let n (+ 1 2 3 4))
5 | ||     (arr n n n)
6 | || "#)?;
  | ||____^ cannot use the `?` operator in a function that returns `()`
7 |  | }
  |  |_- this function should return `Result` or `Option` to accept `?`
  |
  = help: the trait `std::ops::Try` is not implemented for `()`
  = note: required by `std::ops::Try::from_error`

error: aborting due to 4 previous errors

Some errors have detailed explanations: E0277, E0412.
For more information about an error, try `rustc --explain E0277`.
Couldn't compile the test.
---- src/lib.rs - eval (line 97) stdout ----
error: cannot find macro `arr` in this scope
 --> src/lib.rs:99:16
  |
4 | let glsp_arr = arr![60_u8, 120_u16, 240_u32];
  |                ^^^

error: cannot find macro `eval` in this scope
 --> src/lib.rs:101:14
  |
6 | let _: Val = eval!(r#"
  |              ^^^^

error[E0412]: cannot find type `Val` in this scope
 --> src/lib.rs:101:8
  |
6 | let _: Val = eval!(r#"
  |        ^^^ not found in this scope
  |
help: possible candidate is found in another module, you can import it into scope
  |
2 | use glsp::Val;
  |

error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
  --> src/lib.rs:101:14
   |
2  |  / fn main() {
3  |  | let rust_tuple = (60_u8, 120_u16, 240_u32);
4  |  | let glsp_arr = arr![60_u8, 120_u16, 240_u32];
5  |  |
6  |  | let _: Val = eval!(r#"
   |  |______________^
7  | ||     (= [~rust_tuple 2] 480)
8  | ||     (= [~glsp_arr 2] 480)
9  | || "#)?;
   | ||____^ cannot use the `?` operator in a function that returns `()`
...   |
13 |  | assert!(glsp_arr.get::<u32>(2)? == 480);
14 |  | }
   |  |_- this function should return `Result` or `Option` to accept `?`
   |
   = help: the trait `std::ops::Try` is not implemented for `()`
   = note: required by `std::ops::Try::from_error`

error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
  --> src/lib.rs:108:9
   |
2  | / fn main() {
3  | | let rust_tuple = (60_u8, 120_u16, 240_u32);
4  | | let glsp_arr = arr![60_u8, 120_u16, 240_u32];
5  | |
...  |
13 | | assert!(glsp_arr.get::<u32>(2)? == 480);
   | |         ^^^^^^^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `()`
14 | | }
   | |_- this function should return `Result` or `Option` to accept `?`
   |
   = help: the trait `std::ops::Try` is not implemented for `()`
   = note: required by `std::ops::Try::from_error`

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0277, E0412.
For more information about an error, try `rustc --explain E0277`.
Couldn't compile the test.

failures:
    src/lib.rs - compile (line 19)
    src/lib.rs - eval (line 69)
    src/lib.rs - eval (line 82)
    src/lib.rs - eval (line 97)
```

**Comment by fleabitdev** [comment]
Well spotted, thank you - it had honestly slipped my mind that doctests are enabled automatically.

The first item on the roadmap is "add a test suite", so `cargo test` will pass (and actually test the crate, and actually run some valid doctests!) after that item is implemented.

**Comment by fleabitdev** [comment]
The errors have been fixed by 0981514 :)
